swap/src/bitcoin/wallet.rs

```diff
@@ -215,70 +213,117 @@ impl Wallet {
         self.client.lock().await.status_of_script(tx)
     }
 
-    pub async fn watch_until_status<T>(
-        &self,
-        tx: &T,
-        mut status_fn: impl FnMut(ScriptStatus) -> bool,
-    ) -> Result<()>
-    where
-        T: Watchable,
-    {
+    pub async fn subscribe_to(&self, tx: impl Watchable + Send + 'static) -> Subscription {
         let txid = tx.id();
+        self.client
+            .lock()
+            .await
+            .subscriptions
+            .entry(txid)
+            .or_insert({
```

Comment:
This MUST be lazy, otherwise it will evaluate every time this function is called and not just when the entry is absent!

The function to use is [`or_insert_with`](https://doc.rust-lang.org/std/collections/hash_map/enum.Entry.html#method.or_insert_with)!

It might also be worth moving this into a new function so that it is a bit clearer. Eventually, I would also like to de-couple this more from the `tokio` runtime and instead of `spawn`ing the future, just return it to the caller and they have to drive it forward (which likely means spawning).

Reply:
fixed the lazy eval issue. 

I am not entirely sure how returning the future to caller will work so I am leaving it for now. Will it be wrapped in an option to prevent spawning multiple watchers?

Reply:
> I am not entirely sure how returning the future to caller will work so I am leaving it for now. Will it be wrapped in an option to prevent spawning multiple watchers?

Multiple ways to do it I guess. Could also return a dummy future that instantly finishes :)